File: tests/syquex/test_ffi_marshaling.py
"""
FASE 23 — FFI Marshaling Zero-Copy (Manual 4 §7).

TDD: este test ES la especificación del Manual 4 §7.
Si texto_a_c_string no existe en el runtime, el C test NO compila
— eso es correcto. Se corrige el CÓDIGO, no el test.

Manual 4 §7.1: El desafío — Syquex texto (sin \\0) vs C const char* (con \\0)
Manual 4 §7.2: Estrategia zero-copy — añadir byte \\0 al final en la arena
Manual 4 §7.3: texto_a_c_string(CadenaSegura*, Arena*) -> const char*

Comando (Manual 4 §9):
    pytest tests/syquex/test_ffi_marshaling.py -v
Criterio: 0 fugas, 0 copias innecesarias
"""
import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
sys.path.insert(0, PROJECT_ROOT)

BIN_NAME = "test_ffi_marshaling.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)
TEST_SRC = os.path.join(PROJECT_ROOT, "tests", "test_ffi_marshaling.c")

FIXTURE_FFI = os.path.join(
    PROJECT_ROOT, "tests", "fixtures", "test_ffi_marshaling_txt.syq")


def _find_gcc():
    gcc_candidate = os.path.join(PROJECT_ROOT, "toolchain_gcc12", "mingw64", "bin", "gcc.exe")
    if os.path.exists(gcc_candidate):
        return gcc_candidate
    for candidate in ("gcc", "gcc.exe"):
        try:
            subprocess.run([candidate, "--version"], capture_output=True)
            return candidate
        except FileNotFoundError:
            continue
    return gcc_candidate


@pytest.fixture(scope="module")
def exe_path():
    """Compila el test C y retorna la ruta al ejecutable.

    Si texto_a_c_string no existe en el runtime, la compilación FALLA
    — esto es correcto (TDD: el test fuerza la implementación).
    """
    if os.path.exists(BIN_ABS):
        return BIN_ABS

    gcc = _find_gcc()
    r = subprocess.run(
        [gcc, "-O2", "-I", PROJECT_ROOT, "-I.", "-o", BIN_ABS, TEST_SRC,
         os.path.join(PROJECT_ROOT, "synapse_rt_memory.o"),
         "-lm", "-lpthread", "-lws2_32"],
        capture_output=True, text=True, timeout=120
    )
    if r.returncode != 0:
        # TDD: el C no compila porque texto_a_c_string no existe.
        # Esto ES el comportamiento esperado.
        pytest.fail(
            f"test_ffi_marshaling.c NO COMPILA (TDD: falta implementar §7).\n"
            f"Error de gcc:\n{r.stderr[-1000:]}"
        )
    assert os.path.exists(BIN_ABS), f"{BIN_NAME} no se creó tras compilación"
    return BIN_ABS


class TestFFIMarshaling:
    """Manual 4 §7 — FFI Marshaling zero-copy: texto_a_c_string."""

    def test_texto_a_c_string_basico(self, exe_path):
        """§7.2: texto_a_c_string convierte texto Syquex a C string con \\0."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert r.returncode == 0, f"test_ffi_marshaling falló:\n{r.stdout}\n{r.stderr}"
        assert "texto_a_c_string retorna no-NULL" in r.stdout
        assert "strlen(c_str) == 10" in r.stdout
        assert "c_str termina en \\0" in r.stdout
        assert "contenido coincide" in r.stdout

    def test_zero_copy_en_arena(self, exe_path):
        """§7.2: el buffer resultante está en la arena (no en heap separado)."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "c_str NO es el mismo puntero que original" in r.stdout
        assert "c_str está dentro del bloque arena" in r.stdout
        assert "c_str antes del fin de la arena" in r.stdout

    def test_texto_vacio(self, exe_path):
        """§7.2: texto vacío retorna cadena de longitud 0 con \\0."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "texto_a_c_string(vacio) retorna no-NULL" in r.stdout
        assert "cadena vacía termina en \\0 inmediatamente" in r.stdout

    def test_texto_largo(self, exe_path):
        """§7.2: textos largos se convierten correctamente."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "texto_largo retorna no-NULL" in r.stdout
        assert "strlen(texto_largo) == 57" in r.stdout
        assert "texto_largo termina en \\0" in r.stdout

    def test_null_safety(self, exe_path):
        """§7: NULL safety — texto_a_c_string con NULL."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "texto_a_c_string(NULL) retorna NULL" in r.stdout
        assert "texto_a_c_string(_, NULL arena) retorna NULL" in r.stdout

    def test_multiples_conversiones(self, exe_path):
        """§7.2: múltiples conversiones son secuenciales en la arena."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "múltiples conversiones OK" in r.stdout
        assert "s2 > s1" in r.stdout
        assert 's1 == "abc"' in r.stdout
        assert 's2 == "def"' in r.stdout

    def test_no_leaks(self, exe_path):
        """0 fugas: todos los recursos liberados."""
        r = subprocess.run([exe_path], capture_output=True, text=True,
                           encoding='utf-8', errors='replace', timeout=30)
        assert "0 failed" in r.stdout
        assert "RESULTADO" in r.stdout


class TestFFIStringPipeline:
    """ME-3: &T FFI → char* (zero-copy .datos), Manual 3 §9.1/§9.3.

    Valida que el pipeline S1 compila y ejecuta un .syq que usa &texto
    para pasar texto a strlen() via FFI.
    """

    @pytest.fixture(scope="class")
    def exe_path(self, tmp_path_factory):
        from pipeline import ejecutar_compilador
        out = str(tmp_path_factory.mktemp("ffi") / "ffi_txt.exe")
        rc = ejecutar_compilador(FIXTURE_FFI, output_path=out)
        assert rc == 0, f"compilación .syq con &texto rc={rc}"
        assert os.path.exists(out)
        return out

    def test_compila_hasta_exe(self, exe_path):
        assert os.path.getsize(exe_path) > 0

    def test_ffi_strlen_correcto(self, exe_path):
        """strlen(&s) debe devolver 10 para 'Hola Mundo' (Manual 3 §9.3)."""
        e = subprocess.run([exe_path], capture_output=True, text=True,
                           timeout=60, encoding="utf-8", errors="replace")
        assert e.returncode == 0, f"run rc={e.returncode}\n{e.stdout}\n{e.stderr}"
        lineas = [l for l in e.stdout.splitlines() if l.strip()]
        assert lineas == ["10"], f"salida={lineas}"

File: tests/test_ffi_marshaling.c
// FASE 23 — Test de FFI Marshaling Zero-Copy (Manual 4 §7)
// TDD: este test ES la especificación. Si texto_a_c_string no existe,
// el test NO compila — eso es correcto. Se corrige el CÓDIGO, no el test.
//
// Manual 4 §7.1: El desafío — Syquex texto (sin \0) vs C const char* (con \0)
// Manual 4 §7.2: Estrategia zero-copy — añadir byte \0 al final en la arena
// Manual 4 §7.3: texto_a_c_string(CadenaSegura* texto, Arena* arena) -> const char*
//
// Compila: gcc -O2 -I. -I.. -c tests/test_ffi_marshaling.c -o ...
//          gcc -O2 -I. -o test_ffi_marshaling.exe test_ffi_marshaling.o synapse_rt_memory.o -lm -lpthread -lws2_32

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "synapse_rt_types.h"

static int passed = 0;
static int failed = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

int main(void) {
    setbuf(stdout, NULL);

    // ================================================================
    // Manual 4 §7.2: texto_a_c_string — zero-copy conversion
    // ================================================================
    printf("=== 1. texto_a_c_string basico ===\n");
    Arena* arena = arena_crear(4096);

    // Simular un CadenaSegura de Syquex (sin byte nulo)
    const char* original = "Hola Mundo";
    CadenaSegura texto = { .longitud = (int)strlen(original), .datos = original };

    // La función zero-copy debe:
    // 1. Asignar longitud+1 bytes en la arena
    // 2. Copiar los datos
    // 3. Añadir '\0' al final
    // 4. Retornar un puntero al buffer en la arena
    const char* c_str = texto_a_c_string(&texto, arena);
    CHECK(c_str != NULL, "texto_a_c_string retorna no-NULL");
    CHECK(c_str != original, "c_str NO es el mismo puntero que original (copia en arena)");
    CHECK(strlen(c_str) == 10, "strlen(c_str) == 10 (Hola Mundo)");
    CHECK(c_str[10] == '\0', "c_str termina en \\0");
    CHECK(memcmp(c_str, "Hola Mundo", 10) == 0, "contenido coincide");

    // ================================================================
    // Manual 4 §7.2: Verificar zero-copy — el buffer está en la arena
    // ================================================================
    printf("=== 2. Zero-copy: buffer en arena ===\n");
    CHECK((uintptr_t)c_str >= (uintptr_t)arena->inicio,
          "c_str está dentro del bloque arena (no en heap separado)");
    CHECK((uintptr_t)c_str < (uintptr_t)arena->fin,
          "c_str antes del fin de la arena");

    // ================================================================
    // Manual 4 §7.2: Texto vacío
    // ================================================================
    printf("=== 3. Texto vacío ===\n");
    CadenaSegura vacio = { .longitud = 0, .datos = "" };
    const char* c_vacio = texto_a_c_string(&vacio, arena);
    CHECK(c_vacio != NULL, "texto_a_c_string(vacio) retorna no-NULL");
    CHECK(c_vacio[0] == '\0', "cadena vacía termina en \\0 inmediatamente");

    // ================================================================
    // Manual 4 §7.2: Texto largo
    // ================================================================
    printf("=== 4. Texto largo ===\n");
    const char* largo_texto = "Esto es una prueba de marshaling zero-copy para FFI con C";
    CadenaSegura largo = { .longitud = (int)strlen(largo_texto), .datos = largo_texto };
    const char* c_largo = texto_a_c_string(&largo, arena);
    CHECK(c_largo != NULL, "texto_largo retorna no-NULL");
    CHECK(strlen(c_largo) == 57, "strlen(texto_largo) == 57");
    CHECK(c_largo[57] == '\0', "texto_largo termina en \\0");

    // ================================================================
    // Manual 4 §7.2: NULL safety
    // ================================================================
    printf("=== 5. NULL safety ===\n");
    const char* c_null = texto_a_c_string(NULL, arena);
    CHECK(c_null == NULL, "texto_a_c_string(NULL) retorna NULL");

    const char* c_null_arena = texto_a_c_string(&texto, NULL);
    CHECK(c_null_arena == NULL, "texto_a_c_string(_, NULL arena) retorna NULL");

    // ================================================================
    // Manual 4 §7.2: Múltiples conversiones — todas en la arena
    // ================================================================
    printf("=== 6. Multiples conversiones en la arena ===\n");
    const char* s1 = texto_a_c_string(&(CadenaSegura){.longitud=3, .datos="abc"}, arena);
    const char* s2 = texto_a_c_string(&(CadenaSegura){.longitud=3, .datos="def"}, arena);
    CHECK(s1 != NULL && s2 != NULL, "múltiples conversiones OK");
    CHECK(s2 > s1, "s2 > s1 (asignación secuencial en arena)");
    CHECK(strcmp(s1, "abc") == 0, "s1 == \"abc\"");
    CHECK(strcmp(s2, "def") == 0, "s2 == \"def\"");

    arena_free(arena);

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}
